### PR TITLE
feat(design): improve sidebar component

### LINKED ISSUE #277: [FEAT] Add visual hierarchy and bottom nav to sidebar component
# :bulb: Feature request

## Feature Name
- [x] Sidebar Visual Hierarchy  
- [x] Sidebar Bottom Navigation

## The Desired Behavior
Clearly identify hierarchy in the sidebar nav-items and include an option to add a bottom nav to the sidebar to display useful information.

## Your Use Case
As a designer, I'd like to be able to create a visual hierarchy for the nav items in the sidebar and have the ability to include a bottom nav to display additional information.

## Prior Work
N/A

## Environment

<pre><code>
Daffodil version: X.Y.Z
Angular version: X.Y.Z 

 
For Tooling issues:
- Node version: XX  
- Platform:  

Others:

</code></pre>

### LINKED ISSUE #2554: [FEAT] Focus trapping for `over` and `under` sidebar modes
# :bulb: Feature request

## Feature Name
Focus trapping and close on escape for `over` and `under` sidebar modes

## The Desired Behavior
Sidebars with `mode="over"` and `mode="under"` need to handle focus trapping and close on escape. If one of these sidebars open, focus needs to jump to the first focusable element within the sidebar.

This is an important feature for accessibility.

## Your Use Case
As a user, I expect the focus to be within the sidebar once I've opened it.

## Environment

<pre><code>
Daffodil version: X.Y.Z
Angular version: X.Y.Z 

 
For Tooling issues:
- Node version: XX  
- Platform:  

Others:

</code></pre>

### PR DESCRIPTION
**Pull request by xelaint**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/graycoreio/daffodil/blob/develop/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?
1. Sidebar is missing focus trapping. Users currently need to tab through the page to move focus into sidebar.
2. Sidebar is missing custom header and footer components
3. No documentation on sidebar features and usage

Fixes: #277, #2554

## What is the new behavior?
1. Add focus trapping behavior to sidebar so that focus moves to the first focusable element within a sidebar, specifically for `over` and `under` modes.
2. Clean up animation and sidebar visibility so that closed sidebars are not focusable.
3. Create sidebar header that includes a close button that can be hidden, a header title, and custom slotting for additional content.  Add `closeSidebar()` method to header to link closing code to the close button.
4. Create sidebar footer and add body div so that content in body can overflow naturally while allowing footer to stay at bottom of screen.
5. Clean up design land sidebar
6. Extensive documentation on sidebar features and usage

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information

### CODE REVIEW

**Comment by xelaint** [comment]
This PR is dependent on #2570

File: libs/design/src/molecules/sidebar/sidebar/sidebar-theme.scss [hunk 1]
```diff
@@ -3,11 +3,9 @@
 @mixin daff-sidebar-theme($theme) {
 	$base: core.daff-map-deep-get($theme, 'core.base');
 	$base-contrast: core.daff-map-deep-get($theme, 'core.base-contrast');
-	$black: core.daff-map-deep-get($theme, 'core.black');
-	$font-color: core.daff-map-deep-get($theme, 'core.font-color');
 
 	.daff-sidebar {
 		background: $base;
-		color: $font-color;
```
Comment (damienwebdev) [inline libs/design/src/molecules/sidebar/sidebar/sidebar-theme.scss]: Why isn't this `font-color`?

File: libs/design/src/molecules/sidebar/sidebar/sidebar.component.scss [hunk 2]
```diff
@@ -51,4 +51,12 @@
 	&.side-fixed {
 		position: sticky;
 	}
+}
+
+.daff-sidebar {
+	&__body {
+		flex-grow: 1;
+    height: 100%;
```
Comment (damienwebdev) [inline libs/design/src/molecules/sidebar/sidebar/sidebar.component.scss]: Formatting.

File: libs/design/src/molecules/sidebar/sidebar-header/sidebar-header.component.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  Input,
+  ViewEncapsulation,
+} from '@angular/core';
+import { faTimes } from '@fortawesome/free-solid-svg-icons';
+
+@Component({
+  selector: 'daff-sidebar-header',
+  templateUrl: './sidebar-header.component.html',
+  styleUrls: ['./sidebar-header.component.scss'],
+  encapsulation: ViewEncapsulation.None,
```
Comment (damienwebdev) [inline libs/design/src/molecules/sidebar/sidebar-header/sidebar-header.component.ts:18]: Why did you need to set `encapsulation: ViewEncapsulation.None,`?

File: libs/design/sidebar/examples/src/basic-sidebar/basic-sidebar.component.html [hunk 4]
```diff
@@ -1,11 +1,8 @@
...
-    <a daff-link-set-item routerLink="accordion">Accordion</a><br>
-    <a daff-link-set-item routerLink="article">Article</a><br>
+  <daff-sidebar class="basic-sidebar">
+    Sidebar content
   </daff-sidebar>
-  <div>
-    <nav daff-navbar>
-      <div>Basic Sidebar</div>
-    </nav>
+  <div class="basic-sidebar__content">
```
Comment (damienwebdev) [inline libs/design/sidebar/examples/src/basic-sidebar/basic-sidebar.component.html:5]: Why did the content change?

File: apps/design-land/src/app/app.component.scss [hunk 5]
```diff
@@ -3,14 +3,27 @@
...
     margin-left: auto;
   }
 
   &__content {
     padding: 24px 48px 48px;
   }
+
+  &__sidebar {
+    width: 288px;
+    border-right: 1px solid #e8e8e8;
```
Comment (damienwebdev) [inline apps/design-land/src/app/app.component.scss]: We do this border in a few places. Can we embed this in design somehow?

**Review by damienwebdev — CHANGES_REQUESTED** [review]

File: libs/design/src/molecules/sidebar/sidebar/sidebar.component.scss (hunk 2 above)
Comment (xelaint) [reply libs/design/src/molecules/sidebar/sidebar/sidebar.component.scss]: This is just github displaying it weirdly here. Formatting is correct in vscode.

File: libs/design/src/molecules/sidebar/sidebar-header/sidebar-header.component.ts (hunk 3 above)
Comment (xelaint) [reply libs/design/src/molecules/sidebar/sidebar-header/sidebar-header.component.ts:18]: In order to style the title. Otherwise, I'd need to use `ng-deep`.

File: libs/design/sidebar/examples/src/basic-sidebar/basic-sidebar.component.html (hunk 4 above)
Comment (xelaint) [reply libs/design/sidebar/examples/src/basic-sidebar/basic-sidebar.component.html:5]: This is just the example? I just condensed and cleaned it up.

File: apps/design-land/src/app/app.component.scss (hunk 5 above)
Comment (xelaint) [reply apps/design-land/src/app/app.component.scss]: No. This design decision should be up to the end user.

File: libs/design/src/molecules/sidebar/sidebar-viewport/sidebar-viewport.component.ts [hunk 6]
```diff
@@ -133,7 +131,7 @@ export class DaffSidebarViewportComponent implements AfterContentChecked {
   private updateAnimationState() {
     this._animationState = {
       value: getAnimationState(
-        this.sidebars.reduce((acc: boolean, sidebar) => acc || isViewportContentShifted(sidebar.mode, sidebar.open), false),
+        this.sidebars.reduce((acc: boolean, sidebar) => acc || isViewportContentShifted(sidebar.mode, sidebar.open), false), DaffSidebarModeEnum.Over,
```
Comment (damienwebdev) [inline libs/design/src/molecules/sidebar/sidebar-viewport/sidebar-viewport.component.ts]: This needs to be reviewed. I don't remember why we wrote it. It also needs a test.

**Review by damienwebdev — CHANGES_REQUESTED** [review]
## Documentation
- [x] Indicate the optionality of sidebar header and footer in docs.
- [x] Document class props of sidebar header
- [x] Document usage of Sidebar nav in various contexts.
- [x] Figure out the demo of the sidebar nav in different spots.
- [ ] Read docs a few a times.
- [x] Add documentation about importing `BrowserAnimationsModule` or `NoopAnimationsModule` to use sidebar
- [ ] Add documentation of sidebar with anchor scrolling element within body

## Bugs 
- [x] Right sidebar has a horizontal scroll bar in over/under 
- [x] Anchor scrolling on page doesn't work because height is set on `<daff-sidebar-viewport>`. Solution is to set height on `<daff-sidebar-viewport>` only if sidebar is open and mode is `under` or `over`

## API Considerations
- [x] Remove button from header, add slots instead.
- [ ] ~~Replace title with an ngIf~~

## Tests
- [ ] Test focusing code

## Legibility
- [x] Review the clarity of the updateAnimationState in the sidebar viewport

**Comment by xelaint** [comment]
Document example:
- [ ] Sidebar with anchor scrolling

Issue:
Anchor scrolling doesn't work because of height styles in `<daff-sidebar-viewport>`

Possible solution:
Set height on `<daff-sidebar-viewport>` only if sidebar is open and mode is `under` or `over`

File: libs/design/src/molecules/sidebar/sidebar/sidebar-theme.scss (hunk 1 above)
Comment (xelaint) [reply libs/design/src/molecules/sidebar/sidebar/sidebar-theme.scss]: reverted

**Comment by xelaint** [comment]
- [x] Clean up the sidebar viewport animation legibility
- [x] Add an example with sticky content
- [x] Height shift in examples for over and under sidebars
- [x] Make nav a separate animated block to prevent focus from scrolling page back to top
- [x] Make `over` and `under` sidebars `position: fixed`
- [x] Figure out how to stop body scroll when `over` and `under` sidebars are open without interferring with body height

**Comment by xelaint** [comment]
@damienwebdev I've updated `navDesign` to `navPlacement` and added documentation for this. This PR is good to go after you add your portion of the documentation.

In terms of showing the examples for `navPlacement`, the styles we have for a `daff-sidebar-viewport` that is inside of another `daff-sidebar-viewport` is preventing me from adding a working example of `navPlacement="above"` within the sidebar docs.
